#!/usr/bin/env python3
"""
é€šç”¨éŠæˆ²æ¸¬è©¦ä»£ç†ç¨‹å¼ - Sample æ¨¡æ¿
ğŸ”§ AutoTestBuilder æœƒæ ¹æ“šå¯¦éš›éŠæˆ²éœ€æ±‚é€²è¡Œå®¢è£½åŒ–ä¿®æ”¹

æ­¤ç¨‹å¼æ˜¯ä¸€å€‹å®Œå…¨é€šç”¨çš„ AutoTestAgent æ¨¡æ¿ï¼Œå¯ä»¥é©æ‡‰ä»»ä½•åŸºæ–¼ Protocol Buffers çš„éŠæˆ²ã€‚
ä¸»è¦ç‰¹è‰²ï¼š
- å‹•æ…‹ ProtoSchema é©æ‡‰
- æ¨¡çµ„åŒ–ç‹€æ…‹è™•ç†
- å®Œæ•´çš„æ¸¬è©¦æµç¨‹æ”¯æ´
"""

import sys
import os
import time
import signal
import threading
import argparse
from datetime import datetime

# é›™è¼¸å‡ºç³»çµ± - åŒæ™‚è¼¸å‡ºåˆ°æ§åˆ¶å°å’Œæª”æ¡ˆ
class DualOutput:
    def __init__(self, filename):
        self.terminal = sys.stdout
        self.log = open(filename, 'w', encoding='utf-8')
    
    def write(self, message):
        self.terminal.write(message)
        self.log.write(message)
        self.log.flush()
    
    def flush(self):
        self.terminal.flush()
        self.log.flush()
    
    def close(self):
        self.log.close()

# è¨­å®šé›™è¼¸å‡º
dual_output = DualOutput('output.log')
sys.stdout = dual_output
print("ğŸ“ Log è¼¸å‡ºå·²å•Ÿç”¨ï¼Œæª”æ¡ˆ: output.log")

# å‹•æ…‹å°‹æ‰¾ ProtoSchema è·¯å¾‘
proto_loaded = False
proto_paths = [
    os.path.join(os.path.dirname(__file__), '..', 'ProtoSchema'),
    os.path.join(os.path.dirname(__file__), 'ProtoSchema'),
    '/mnt/d/AutoTest/ProtoSchema'
]

for proto_path in proto_paths:
    if os.path.exists(proto_path):
        sys.path.append(proto_path)
        print(f"âœ“ æ‰¾åˆ° ProtoSchema è·¯å¾‘: {os.path.abspath(proto_path)}")
        proto_loaded = True
        break

if not proto_loaded:
    print("âŒ æ‰¾ä¸åˆ° ProtoSchema è·¯å¾‘")
    sys.exit(1)

# å‹•æ…‹å°å…¥ Proto æ¨¡çµ„
try:
    import InputCommand_pb2 as InputCommand
    import GameFlowData_pb2 as GameFlowData
    print("âœ“ Proto modules loaded successfully")
except ImportError as e:
    print(f"âŒ Failed to import proto modules: {e}")
    sys.exit(1)

# å°å…¥å…¶ä»–æ¨¡çµ„
from core.udp_manager import UDPManager as UDPSocketManager
from core.message_handler import MessageHandler
from core.statistics_manager import StatisticsManager
from flow.state_manager import StateManager
from config.target_parser import TargetParser

class CustomGameTestAgent:
    """é€šç”¨éŠæˆ²æ¸¬è©¦ä»£ç†ç¨‹å¼ - Sample æ¨¡æ¿
    ğŸ”§ AutoTestBuilder æœƒæ ¹æ“šå¯¦éš›éŠæˆ²éœ€æ±‚é€²è¡Œå®¢è£½åŒ–ä¿®æ”¹
    """
    
    def __init__(self, requirement: str = "é€šç”¨è‡ªå‹•æ¸¬è©¦"):
        self.requirement = requirement
        self.running = False
        
        # åˆå§‹åŒ–æ ¸å¿ƒå…ƒä»¶
        self._init_components()
        
        # è¨­å®šä¿¡è™Ÿè™•ç†
        signal.signal(signal.SIGINT, self._signal_handler)
        signal.signal(signal.SIGTERM, self._signal_handler)
        
        print("ğŸ¤– é€šç”¨éŠæˆ²æ¸¬è©¦ä»£ç†ç¨‹å¼å·²åˆå§‹åŒ– (Dynamic ProtoSchema Version)")
        print(f"ğŸ¯ æ¸¬è©¦éœ€æ±‚: {self.requirement}")
        print("ğŸ“¡ ç›®æ¨™ä¼ºæœå™¨: UDP Server")
    
    def _init_components(self):
        """åˆå§‹åŒ–æ ¸å¿ƒå…ƒä»¶"""
        # UDP ç®¡ç†å™¨
        self.udp_manager = UDPSocketManager(
            server_url="udp://127.0.0.1:8587"
        )
        
        # è¨Šæ¯è™•ç†å™¨
        self.message_handler = MessageHandler()
        
        # çµ±è¨ˆç®¡ç†å™¨
        self.statistics_manager = StatisticsManager()
        
        # ç‹€æ…‹ç®¡ç†å™¨
        self.state_manager = StateManager()
        
        # ç›®æ¨™è§£æå™¨
        self.target_parser = TargetParser()
        
        print("âœ“ æ‰€æœ‰æ ¸å¿ƒå…ƒä»¶åˆå§‹åŒ–å®Œæˆ")
    
    def start(self):
        """å•Ÿå‹•æ¸¬è©¦ä»£ç†ç¨‹å¼"""
        print("ğŸš€ å•Ÿå‹•éŠæˆ²æ¸¬è©¦ä»£ç†ç¨‹å¼...")
        
        # è§£ææ¸¬è©¦éœ€æ±‚
        targets = self.target_parser.parse_requirement(self.requirement)
        if targets:
            print(f"ğŸ¯ è§£æéœ€æ±‚æˆåŠŸ: {self.target_parser.get_target_summary()}")
            self.state_manager.set_targets(targets)
        
        # å»ºç«‹ UDP é€£ç·š
        self.udp_manager.connect(
            on_open=self._on_open,
            on_message=self._on_message,
            on_error=self._on_error,
            on_close=self._on_close
        )
    
    def stop(self):
        """åœæ­¢æ¸¬è©¦ä»£ç†ç¨‹å¼"""
        print("ğŸ›‘ åœæ­¢éŠæˆ²æ¸¬è©¦ä»£ç†ç¨‹å¼...")
        self.running = False
        if hasattr(self, 'udp_manager'):
            self.udp_manager.close()
    
    def _on_open(self, ws):
        """UDP é€£ç·šå»ºç«‹å›èª¿"""
        print("âœ… UDP é€£ç·šå·²å»ºç«‹")
        self.running = True
        self.statistics_manager.start_session()
        
        # å•Ÿå‹•çµ±è¨ˆå ±å‘Šç·šç¨‹
        def print_stats():
            while self.running:
                time.sleep(10)  # æ¯10ç§’è¼¸å‡ºä¸€æ¬¡çµ±è¨ˆ
                if self.running:
                    self.statistics_manager.print_statistics()
        
        stats_thread = threading.Thread(target=print_stats, daemon=True)
        stats_thread.start()
    
    def _on_message(self, ws, message):
        """è™•ç†æ¥æ”¶åˆ°çš„è¨Šæ¯"""
        try:
            # è§£æéŠæˆ²æ•¸æ“š
            game_data = self.message_handler.parse_message(message)
            if game_data:
                # æ›´æ–°ç‹€æ…‹ç®¡ç†å™¨ç‹€æ…‹
                self.state_manager.update_game_state(game_data)
                
                # ç”Ÿæˆä¸¦ç™¼é€è¼¸å…¥
                input_command = self.state_manager.generate_input()
                if input_command:
                    # åœ¨å¯¦éš›ç™¼é€æ™‚å°å‡º log
                    self._log_input_command(input_command)
                    
                    # æ·»åŠ ç™¼é€å‰çš„ç¢ºèªä¿¡æ¯
                    print(f"ğŸš€ æº–å‚™ç™¼é€è¼¸å…¥æŒ‡ä»¤åˆ° UDP...")
                    
                    self.udp_manager.send_message(input_command)
                    self.statistics_manager.record_input_sent()
                    
                    print(f"âœ… è¼¸å…¥æŒ‡ä»¤ç™¼é€å®Œæˆ")
                else:
                    # å¦‚æœæ²’æœ‰ç”Ÿæˆè¼¸å…¥ï¼Œä¹Ÿè¦è¨˜éŒ„
                    print(f"â¸ï¸  æœ¬æ¬¡å¾ªç’°æ²’æœ‰ç”Ÿæˆè¼¸å…¥æŒ‡ä»¤")
                    
            # æ›´æ–°çµ±è¨ˆ
            self.statistics_manager.record_message_received()
            
        except Exception as e:
            print(f"âŒ Error processing message: {e}")
    
    def _on_error(self, ws, error):
        """è™•ç†é€£ç·šéŒ¯èª¤"""
        print(f"âŒ UDP éŒ¯èª¤: {error}")
    
    def _on_close(self, ws, close_status_code, close_msg):
        """è™•ç†é€£ç·šé—œé–‰"""
        print(f"ğŸ”Œ UDP é€£ç·šå·²é—œé–‰ (ç‹€æ…‹ç¢¼: {close_status_code})")
        self.running = False
        self.statistics_manager.end_session()
        self.state_manager.stop()
    
    def _signal_handler(self, signum, frame):
        """è™•ç†ç³»çµ±ä¿¡è™Ÿ"""
        print(f"\nğŸ›‘ æ¥æ”¶åˆ°åœæ­¢ä¿¡è™Ÿ ({signum})")
        self.stop()
        dual_output.close()
        sys.exit(0)

    def _log_input_command(self, input_command: bytes):
        """è§£æä¸¦å°å‡ºè¼¸å…¥æŒ‡ä»¤çš„ log"""
        try:
            # TODO: ğŸ”§ éœ€è¦è§£æ InputCommand_pb2 çµæ§‹ - ä¾†æº: AutoTest_Game_Setting.md
            # å˜—è©¦è§£æè¼¸å…¥æŒ‡ä»¤
            input_cmd = InputCommand.InputCommand()
            input_cmd.ParseFromString(input_command)
            
            # ç²å–æŒ‰éµåç¨±
            key_names = []
            for key in input_cmd.key_inputs:
                key_name = self._get_key_name(key)
                key_names.append(key_name)
            
            if key_names:
                action = "æŒ‰ä¸‹" if input_cmd.is_key_down else "é‡‹æ”¾"
                print(f"ğŸ“¤ ç™¼é€{action}è¼¸å…¥æŒ‡ä»¤: [{', '.join(key_names)}]")
            
        except Exception as e:
            print(f"ğŸ“¤ ç™¼é€è¼¸å…¥æŒ‡ä»¤ ({len(input_command)} bytes)")
    
    def _get_key_name(self, key_value: int) -> str:
        """ç²å–æŒ‰éµåç¨±"""
        # TODO: ğŸ”§ éœ€è¦å¾ AutoTest_Game_Setting.md çš„ EInputKeyType æšèˆ‰å–å¾—æŒ‰éµåç¨±
        key_names = {
            0: "UP",    # TODO: ğŸ”§ å°æ‡‰ EInputKeyType::Up
            1: "DOWN",  # TODO: ğŸ”§ å°æ‡‰ EInputKeyType::Down  
            2: "LEFT",  # TODO: ğŸ”§ å°æ‡‰ EInputKeyType::Left
            3: "RIGHT", # TODO: ğŸ”§ å°æ‡‰ EInputKeyType::Right
            4: "START", # TODO: ğŸ”§ å°æ‡‰ EInputKeyType::Start
            5: "NITRO", # TODO: ğŸ”§ å°æ‡‰ EInputKeyType::Nitro
            8: "COIN",  # TODO: ğŸ”§ å°æ‡‰ EInputKeyType::Coin
        }
        return key_names.get(key_value, f"UNKNOWN_KEY_{key_value}")

def main():
    """ä¸»ç¨‹å¼å…¥å£"""
    parser = argparse.ArgumentParser(description='é€šç”¨éŠæˆ²æ¸¬è©¦ä»£ç†ç¨‹å¼ - Sample æ¨¡æ¿')
    parser.add_argument('--requirement', '-r',
                       default="é€šç”¨è‡ªå‹•æ¸¬è©¦",
                       help='æ¸¬è©¦éœ€æ±‚æè¿°')
    parser.add_argument('--url', '-u',
                       default="udp://127.0.0.1:8587",
                       help='Server URL (UDP Socket connection)')
    
    args = parser.parse_args()
    
    try:
        # å‰µå»ºä¸¦å•Ÿå‹•æ¸¬è©¦ä»£ç†ç¨‹å¼
        agent = CustomGameTestAgent(requirement=args.requirement)
        agent.start()
        
        # ä¿æŒç¨‹å¼é‹è¡Œ
        while agent.running:
            time.sleep(1)
            
    except KeyboardInterrupt:
        print("\nğŸ›‘ ä½¿ç”¨è€…ä¸­æ–·ç¨‹å¼")
    except Exception as e:
        print(f"âŒ ç¨‹å¼åŸ·è¡ŒéŒ¯èª¤: {e}")
    finally:
        if 'agent' in locals():
            agent.stop()
        dual_output.close()

if __name__ == "__main__":
    main()
